File: solutions/go/binary-search/1/binary_search.go
package binarysearch

func SearchInts(list []int, key int) int {
	left := 0
	right := len(list) - 1
	var middle, guess int

	for left <= right {
		middle = int((left + right) / 2)
		guess = list[middle]

		if guess == key {
			return middle
		} else if guess < key {
			left = middle + 1
		} else if guess > key {
			right = middle - 1
		}
	}

	return -1
}

File: solutions/clojure/binary-search/1/src/binary_search.clj
(ns binary-search)

(defn search-for [target seq]
  (loop [left 0
         right (dec (count seq))]
    (if (> left right)
      (throw (new Exception "not found")))
    (let [middle (int (/ (+ left right) 2))
          found (nth seq middle)]
      (cond
        (< found target) (recur (inc middle) right)
        (> found target) (recur left (dec middle))
        (= found target) middle))))

(defn middle [seq] ;; not used
  (int (/ (count seq) 2)))